// @atlaskit/onboarding/dist/esm/components/SpotlightTransition.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { createContext } from 'react';
var TransitionContext = /*#__PURE__*/createContext({
  isOpen: true,
  onExited: function onExited() {}
}); // checks if children exist and are truthy

var hasChildren = function hasChildren(children) {
  return React.Children.count(children) > 0 && React.Children.map(children, function (child) {
    return !!child;
  }).filter(Boolean).length > 0;
};

var Transition = /*#__PURE__*/function (_React$Component) {
  _inherits(Transition, _React$Component);

  var _super = _createSuper(Transition);

  function Transition() {
    var _this;

    _classCallCheck(this, Transition);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      currentChildren: undefined
    });

    _defineProperty(_assertThisInitialized(_this), "onExited", function () {
      _this.setState({
        currentChildren: _this.props.children
      });
    });

    return _this;
  }

  _createClass(Transition, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(TransitionContext.Provider, {
        value: {
          onExited: this.onExited,
          isOpen: hasChildren(this.props.children)
        }
      }, this.state.currentChildren);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      var previousChildren = state.currentChildren;
      var exiting = hasChildren(previousChildren) && !hasChildren(props.children);
      return {
        currentChildren: exiting ? previousChildren : props.children
      };
    }
  }]);

  return Transition;
}(React.Component);

export var SpotlightTransitionConsumer = TransitionContext.Consumer;
export default Transition;
