// @atlassian/switcher/dist/esm/types.js
export var RecentContainerType;

(function (RecentContainerType) {
  RecentContainerType["JIRA_PROJECT"] = "jira-project";
  RecentContainerType["CONFLUENCE_SPACE"] = "confluence-space";
})(RecentContainerType || (RecentContainerType = {}));

export var CollaborationGraphRecentContainerType;

(function (CollaborationGraphRecentContainerType) {
  CollaborationGraphRecentContainerType["JIRA_PROJECT"] = "jiraProject";
  CollaborationGraphRecentContainerType["CONFLUENCE_SPACE"] = "confluenceSpace";
})(CollaborationGraphRecentContainerType || (CollaborationGraphRecentContainerType = {}));

export var Permissions;

(function (Permissions) {
  Permissions["MANAGE"] = "manage";
  Permissions["CAN_INVITE_USERS"] = "invite-users";
})(Permissions || (Permissions = {}));

export var Product;

(function (Product) {
  Product["BITBUCKET"] = "bitbucket";
  Product["CONFLUENCE"] = "confluence";
  Product["HOME"] = "home";
  Product["JIRA"] = "jira";
  Product["SITE_ADMIN"] = "site-admin";
  Product["TRELLO"] = "trello";
  Product["START"] = "start";
})(Product || (Product = {}));

export var Feature;

(function (Feature) {
  Feature["disableCustomLinks"] = "disableCustomLinks";
  Feature["enableRecentContainers"] = "enableRecentContainers";
  Feature["enableRemoteConfiguration"] = "enableRemoteConfiguration";
  Feature["enableRemoteAdminLinks"] = "enableRemoteAdminLinks";
  Feature["disableSwitchToHeading"] = "disableSwitchToHeading";
  Feature["xflow"] = "xflow";
  Feature["isProductStoreInTrelloJSWFirstEnabled"] = "isProductStoreInTrelloJSWFirstEnabled";
  Feature["isProductStoreInTrelloConfluenceFirstEnabled"] = "isProductStoreInTrelloConfluenceFirstEnabled";
  Feature["isSlackDiscoveryEnabled"] = "isSlackDiscoveryEnabled";
})(Feature || (Feature = {}));

export var AnalyticsItemType;

(function (AnalyticsItemType) {
  AnalyticsItemType["PRODUCT"] = "product";
  AnalyticsItemType["ADMIN"] = "admin";
  AnalyticsItemType["TRY"] = "try";
  AnalyticsItemType["JOIN"] = "join";
  AnalyticsItemType["CUSTOM_LINK"] = "customLink";
  AnalyticsItemType["DISCOVER_FIXED_LINKS"] = "discover-fixed-links";
  AnalyticsItemType["RECENT"] = "recent";
})(AnalyticsItemType || (AnalyticsItemType = {}));

export var ProductKey;

(function (ProductKey) {
  ProductKey["CONFLUENCE"] = "confluence.ondemand";
  ProductKey["JIRA_CORE"] = "jira-core.ondemand";
  ProductKey["JIRA_SOFTWARE"] = "jira-software.ondemand";
  ProductKey["JIRA_SERVICE_DESK"] = "jira-servicedesk.ondemand";
  ProductKey["BITBUCKET"] = "bitbucket";
  ProductKey["OPSGENIE"] = "opsgenie";
  ProductKey["STATUSPAGE"] = "statuspage";
  ProductKey["TRELLO"] = "trello";
  ProductKey["COMPASS"] = "compass";
  ProductKey["TEAM_CENTRAL"] = "townsquare";
  ProductKey["AVOCADO"] = "avocado";
})(ProductKey || (ProductKey = {}));

export var DiscoverLinkItemKeys;

(function (DiscoverLinkItemKeys) {
  DiscoverLinkItemKeys["DISCOVER_MORE"] = "discover-more";
  DiscoverLinkItemKeys["SLACK_INTEGRATION"] = "slack-integration";
})(DiscoverLinkItemKeys || (DiscoverLinkItemKeys = {}));
